### PR TITLE
Use a wrapper to detect multiplicity coercions

### PR DESCRIPTION
**Pull request by monoidal**
Refs #354

### CODE REVIEW

File: compiler/deSugar/DsBinds.hs [hunk 1]
```diff
@@ -1140,7 +1140,9 @@ dsHsWrapper (WpCast co)       = ASSERT(coercionRole co == Representational)
                                 return $ \e -> mkCastDs e co
 dsHsWrapper (WpEvApp tm)      = do { core_tm <- dsEvTerm tm
                                    ; return (\e -> App e core_tm) }
-
+dsHsWrapper (WpMultCoercion co) = do { when (not (isReflCo co)) $
```
Comment (aspiwack) [inline compiler/deSugar/DsBinds.hs]: As as side remark, there may be coercions which are functionally reflexivity but where `isReflCo` returns false, I believe. Typically, the transitivity rule applies to two reflexivity. We may want, at some point to be resilient to that.

I'm just raising this point for the sake of awareness.

File: compiler/typecheck/TcEnv.hs [hunk 2]
```diff
@@ -548,36 +550,34 @@ tcExtendIdEnv :: [TcId] -> TcM a -> TcM a
...
-tcExtendIdEnv1 name (Scaled id_mult id) thing_inside
-  = do { (local_usage, result) <- tcCollectingUsage $ tcExtendIdEnv2 [(name,id)] thing_inside
-       ; check_then_add_usage local_usage
-       ; return result }
+tcExtendIdEnv1 :: Name -> TcId -> TcM a -> TcM a
+-- Exactly like tcExtendIdEnv2, but for a single (name,id) pair
+tcExtendIdEnv1 name id thing_inside
+  = tcExtendIdEnv2 [(name,id)] thing_inside
+
+tcExtendIdEnv1Scaled :: Name -> Scaled TcId -> TcM a -> TcM (a, HsWrapper)
```
Comment (aspiwack) [inline compiler/typecheck/TcEnv.hs]: Simon suggested, instead, to have a `checkUsage` function or such, and simply call `checkUsage . tcExtendIdEnv1` where relevant.

File: compiler/typecheck/TcExpr.hs [hunk 3]
```diff
@@ -379,7 +379,7 @@ tcExpr expr@(OpApp fix arg1 op arg2) res_ty
        ; (wrap_arg1, [arg2_sigma], op_res_ty) <-
            matchActualFunTys doc orig1 (Just (unLoc arg1)) 1 arg1_ty
 
-       ; tcSubMult AppOrigin Omega (scaledMult arg2_sigma)
+       ; wrapper <- tcSubMult AppOrigin Omega (scaledMult arg2_sigma)
```
Comment (aspiwack) [inline compiler/typecheck/TcExpr.hs]:
```suggestion
       ; mult_wrapper <- tcSubMult AppOrigin Omega (scaledMult arg2_sigma)
```

File: compiler/typecheck/TcExpr.hs [hunk 4]
```diff
@@ -424,7 +424,7 @@ tcExpr expr@(OpApp fix arg1 op arg2) res_ty
              --
              -- We need to zonk here as well, see Dollar2 for an example
              wrap1 = mkWpFun idHsWrapper wrap_res arg2_sigma res_ty doc
-                     <.> wrap_arg1
+                     <.> wrap_arg1 <.> wrapper
```
Comment (aspiwack) [inline compiler/typecheck/TcExpr.hs]:
```suggestion
                     <.> wrap_arg1 <.> mutl_wrapper
```

**Comment by monoidal** [comment]
Back for review. The build passes: https://circleci.com/gh/tweag/ghc/4674

**Comment by monoidal** [comment]
Let's not merge yet: https://gitlab.haskell.org/ghc/ghc/merge_requests/1438#note_216701

File: compiler/deSugar/DsBinds.hs (hunk 1 above)
Comment (goldfirere) [reply compiler/deSugar/DsBinds.hs]: Yes -- use `isReflexiveCo` instead.

**Pull request merged**